ouroboros-consensus-shelley/test/Test/ThreadNet/TxGen/Shelley.hs

```diff
@@ -1,14 +1,113 @@
+{-# LANGUAGE DataKinds                #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE FlexibleInstances        #-}
+{-# LANGUAGE NamedFieldPuns           #-}
+{-# LANGUAGE TypeFamilies             #-}
+{-# LANGUAGE ViewPatterns             #-}
 {-# OPTIONS_GHC -Wno-orphans #-}
-module Test.ThreadNet.TxGen.Shelley () where
+module Test.ThreadNet.TxGen.Shelley
+  ( ShelleyTxGenExtra(..)
+  ) where
+
+import           Control.Monad.Except (runExcept)
+import           Crypto.Number.Generate (generateBetween, generateMax)
+import           Crypto.Random (MonadRandom)
+import qualified Data.Sequence as Seq
+
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Ledger.Abstract
+
+import qualified Shelley.Spec.Ledger.LedgerState as SL
+import qualified Shelley.Spec.Ledger.STS.Ledger as STS
+import qualified Shelley.Spec.Ledger.Tx as STx
+
+import           Ouroboros.Consensus.Shelley.Ledger
+
+import           Test.QuickCheck.Gen (Gen (..), suchThat)
+import           Test.QuickCheck.Random (mkQCGen)
 
 import           Test.ThreadNet.TxGen (TxGen (..))
 
-import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock)
+import qualified Test.Shelley.Spec.Ledger.ConcreteCryptoTypes as CSL
+import qualified Test.Shelley.Spec.Ledger.Generator.Core as Gen
+import qualified Test.Shelley.Spec.Ledger.Generator.Utxo as Gen
 
 import           Test.Consensus.Shelley.MockCrypto (TPraosMockCrypto)
 
+data ShelleyTxGenExtra = ShelleyTxGenExtra
+  { -- | Generator environment.
+    stgeGenEnv :: Gen.GenEnv
+  }
 
 instance TxGen (ShelleyBlock TPraosMockCrypto) where
+
+  type TxGenExtra (ShelleyBlock TPraosMockCrypto) = ShelleyTxGenExtra
+
   -- TODO #1823
-  testGenTxs _ _ _ _ = return []
+  testGenTxs _numCoreNodes curSlotNo cfg lst stge = do
+      n <- generateBetween 0 20
+      go [] n $ applyChainTick (configLedger cfg) curSlotNo lst
+    where
+      go :: MonadRandom m
+         => [GenTx (ShelleyBlock TPraosMockCrypto)]  -- ^ Accumulator
+         -> Integer  -- ^ Number of txs to still produce
+         -> TickedLedgerState (ShelleyBlock TPraosMockCrypto)
+         -> m [GenTx (ShelleyBlock TPraosMockCrypto)]
+      go acc 0 _  = return (reverse acc)
+      go acc n st = do
+        tx <- quickCheckAdapter $ genTx cfg st (stgeGenEnv stge)
+        case runExcept $ applyTx (configLedger cfg) tx st of
+          -- We don't mind generating invalid transactions
+          Left  _   -> go (tx:acc) (n - 1) st
+          Right st' -> go (tx:acc) (n - 1) st'
+
+genTx
+  :: TopLevelConfig (ShelleyBlock TPraosMockCrypto)
+  -> TickedLedgerState (ShelleyBlock TPraosMockCrypto)
+  -> Gen.GenEnv
+  -> Gen (GenTx (ShelleyBlock TPraosMockCrypto))
+genTx _cfg TickedLedgerState { tickedSlotNo, tickedLedgerState } genEnv =
+    mkShelleyTx <$> Gen.genTx
+      genEnv
+      ledgerEnv
+      (utxoSt, dpState) `suchThat` simpleTx
+  where
+    simpleTx (STx._body -> txb) =
+      (Seq.null $ STx._certs txb)
```

Comment:
The transaction generator depends only on the chain state it's given, right? So just discarding the delegation certs it generates (and assumes succeed, if I understand correctly) won't cause it to generate invalid transactions for later slots?

Reply:
Should we also be discarding generated proposal transactions?